flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/PipelinedRegionSchedulingStrategy.java

```diff
@@ -0,0 +1,166 @@
+
+		this.schedulerOperations = checkNotNull(schedulerOperations);
+		this.schedulingTopology = checkNotNull(schedulingTopology);
+
+		init();
+	}
+
+	private void init() {
+		for (SchedulingPipelinedRegion<?, ?> region : schedulingTopology.getAllPipelinedRegions()) {
+			for (SchedulingResultPartition<?, ?> partition : region.getConsumedResults()) {
+				// sanity check
+				checkState(partition.getResultType() == ResultPartitionType.BLOCKING);
+
+				partitionConsumerRegions.computeIfAbsent(partition.getId(), pid -> new HashSet<>()).add(region);
+				correlatedResultPartitions.computeIfAbsent(partition.getResultId(), rid -> new HashSet<>()).add(partition);
+			}
+		}
+	}
+
+	@Override
+	public void startScheduling() {
+		final Set<SchedulingPipelinedRegion<?, ?>> sourceRegions = IterableUtils
+			.toStream(schedulingTopology.getAllPipelinedRegions())
+			.filter(region -> !region.getConsumedResults().iterator().hasNext())
+			.collect(Collectors.toSet());
+		maybeScheduleRegions(sourceRegions);
+	}
+
+	@Override
+	public void restartTasks(final Set<ExecutionVertexID> verticesToRestart) {
+		final Set<SchedulingPipelinedRegion<?, ?>> regionsToRestart = verticesToRestart.stream()
+			.map(schedulingTopology::getPipelinedRegionOfVertex)
+			.collect(Collectors.toSet());
+		maybeScheduleRegions(regionsToRestart);
+	}
+
+	@Override
+	public void onExecutionStateChange(final ExecutionVertexID executionVertexId, final ExecutionState executionState) {
+		if (executionState == ExecutionState.FINISHED) {
+			final Set<SchedulingResultPartition<?, ?>> finishedPartitions = IterableUtils
+				.toStream(schedulingTopology.getVertex(executionVertexId).getProducedResults())
+				.filter(partition -> partitionConsumerRegions.containsKey(partition.getId()))
+				.filter(partition -> partition.getState() == ResultPartitionState.CONSUMABLE)
+				.flatMap(partition -> correlatedResultPartitions.get(partition.getResultId()).stream())
+				.collect(Collectors.toSet());
+
+			final Set<SchedulingPipelinedRegion<?, ?>> consumerRegions = finishedPartitions.stream()
+				.flatMap(partition -> partitionConsumerRegions.get(partition.getId()).stream())
+				.collect(Collectors.toSet());
+			maybeScheduleRegions(consumerRegions);
+		}
+	}
+
+	@Override
+	public void onPartitionConsumable(final IntermediateResultPartitionID resultPartitionId) {
+	}
+
+	private void maybeScheduleRegions(final Set<SchedulingPipelinedRegion<?, ?>> regions) {
+		final List<SchedulingPipelinedRegion<?, ?>> regionsSorted =
+			SchedulingStrategyUtils.sortPipelinedRegionsInTopologicalOrder(schedulingTopology, regions);
+		for (SchedulingPipelinedRegion<?, ?> region : regionsSorted) {
+			maybeScheduleRegion(region);
+		}
+	}
+
+	private void maybeScheduleRegion(final SchedulingPipelinedRegion<?, ?> region) {
+		if (!areRegionInputsAllConsumable(region)) {
+			return;
+		}
+
+		// sanity check
+		if (!areRegionVerticesAllInCreatedState(region)) {
```

Comment:
Maybe `checkState` for consistency, and drop the `// sanity check` comment.

Reply:
ok.